### DISCUSSION TITLE
How to make JSBSim interface with outside running simulator?

### DISCUSSION THREAD

**Discussion by GloryENG** [JSBSim-Team/jsbsim#785] [Q&A]
Dear reader and developter...
In advance, thanks for the developers about giving this precious simulation evironments.

Thesedays, I have tried to make interface between JSBSim and outside simulator, replacing flightgear.

the 'outside simulator' is that something like flighgear or UE5, but not excatly. I am going to design some simple flight simulator using Unity.

In this point, the code for interfacing between JSBSim and Unity simulator is required.

First, where can I find the list of parameters that simulator user can input to simulator, and can get output from the simulator?
As you know, there are variety of user inputs not only throttle, roll, yaw, pitch, but also landing gear retraction, flap...like pilot can control in the cockpit pannel.....Also, I know that the user can control some kinds of weather effects....
Where  can i find these kinds of user controlable parameters to send and interface with JSBSIM?
In the same way, JSBSim generates lots of parameters not only flight states but also variety of simulation results. How can i check the list of simulation output parameter lists?

Next, if i could define the input and output (O/I) parameters for simulation, How can i get and set these paremeters to JSBSim from outside software(standalone). I saw 'copy_to_JSBSim() and copy_from_JSBSim() in the Flightgear.cxx code, but I don't have definite idea how to control the parameters inside JSBSim using these command...

Actually, I am not English native speaker, so my questions could be awkward, but your kind answer will be so much helpful to me.

Thanks in advance.
@bcoconni

**Comment by GloryENG** [comment]
Hello, I am the writer.
Before writting this post, I wrote the article below:
https://github.com/JSBSim-Team/jsbsim/issues/228

In the comment by 'bcoconni'on Nov 1, 2019
there are some examples of code as below:
   FCS->SetDaCmd( globals->get_controls()->get_aileron()); 
   FCS->SetRollTrimCmd( globals->get_controls()->get_aileron_trim() ); 
or
  _set_Euler_Angles( Propagate->GetEuler(FGJSBBase::ePhi), 
                        Propagate->GetEuler(FGJSBBase::eTht), 
                        Propagate->GetEuler(FGJSBBase::ePsi) ); 

In this point, there must be some parameter lists which can be get or set by these command. I am wondering this parameter lists and the exact variable names to access in the code.

**Reply by GloryENG** [reply]
Now, I found that two things related to above question,

1. some aircraft xml file includes their input and output parameters at the end part of it.  for example, at the last part of 737.xlm file, 41 output properties, 6 input properties, and its port number are defined. I think that it is so called QTJSBsim? However, not every aircraft xlm include these kind of definition.....

2. in the directory of jsbsim/data-output, I can find what the parameters JSBSim generate and output through the simulation.

However, I can not figure out the whole accessible input/output parameters, properties for performing simulation, and how to control/access(get and set), still...

please help me~

**Comment by dpculp** [comment]
Hello @GloryENG , I'm not sure what your goal is here.  Is it like QtJSBSim?  [https://sourceforge.net/projects/qtjsbsim/](url).   QtJSBSim shows how to set up bidirectional communication with an instance of JSBSim Standalone app using UDP sockets.

**Reply by GloryENG** [reply]
Thanks for your kind answer. As I said, My mother language is not English, so my questions and sentences could be hard to understand....

First of all, my final goal is to develope simulators which computes its dynamics with JSBSim while visualizes aircrafts flying states with my own software, instead of flightgear.

To do this, I need to communicate between my visualizing software and JSBSim. 
From my visualizing software, the user or operator gives pilot inputs, like roll/pitch/yaw/throttle and landing gear command, the it sends this command to JSBsim. Then, JSBSim computes the FDM and generate states and properties, which are sent to my visualizing software and shown.

For this, as @seanmcleod said below, I need to write the code regarding copy_to_JSBSim and copy_from_JSBSIm. 
In this point, I need to know what properties, states, and values I can access through copy_to_JSBSim and copy_from_JSBSIm, and how to....

**Reply by eeeewwqq** [reply]
`JSBSim.exe --aircraft=737 --catalog`

you can see all properties from tree for selected model with this cli command

**Comment by seanmcleod** [comment]
@GloryENG if you want to implement a completely generic flight simulator then your best bet is probably to take a look at the FlightGear source code to see how they interface to JSBSim.

I've implemented a custom flight simulator using JSBSim, where I use Prepar3D to generate the external view across 3 LCDs/projectors and so I've written a small amount of glue code in `copy_to_JSBSim()` which reads the physical inceptors (joystick, throttle, rudder pedals) and passes those into JSBSim using the `FCS` class and then in `copy_from_JSBSim()` I have code that queries the current state in terms of 3D position, attitude etc. and uses that data to update Prepar3D so that it can update it's visual output.

In my case I'm not trying to support any and all aircraft so I only support a fixed/known set of inceptors and only landing gear position and flap position.

See this comment for an example code snippet - https://github.com/JSBSim-Team/jsbsim/discussions/773#discussioncomment-4252663

In general JSBSim has a property tree where some items are read-only and some items are readable and writeable. Some of these properties also have explicit class methods e.g. `FCS->SetDeCmd()` that can be used optionally, or you can simply use the property tree methods to read+write properties. A number of these properties come from JSBSim, i.e. hard coded in the JSBSim source but others can be defined by the aircraft writer by creating new properties in the aircraft FDM file.

**Reply by GloryENG** [reply]
Thanks for your kind answer @seanmcleod .

What i am going to do is exactly same thing that you mentioned. I agree with your mention that the best and easiest why to do this is to use copy_to/from_JSBsim in FlightGear source code...

In this point, I think that the user inputs required for each aircraft could be different, e.g. some aircraft has spoiler, flaps, external lights but others do not. where can I find the defined inputs for the aircraft? and how to access them?

Furthermore, to do the flight simulation based on the designed scenarios, user must send some environmental information and initial condition to JSBSim, and how can I send this properties to JSBsim?

In addition, I am wondering what "property tree method" in your answer. 

I have had lots of times to get information regarding above questions, but I knowledge is to short too catch these information..

**Comment by seanmcleod** [comment]
> and how to access them?
> I am wondering what "property tree method"

Via the property tree manager, e.g. see some code snippets in this comment showing the use of the property manager - https://github.com/JSBSim-Team/jsbsim/discussions/627#discussioncomment-2817930

But yes in general you need some mapping between the user's inputs, e.g. hitting the `L` key on the keyboard for example is mapped to lowering the landing gear, and what the property path is for that particular property, assuming an aircraft has retractable gear.

And handling differences e.g. between piston aircraft which include say a fuel mixture control and turbine aircraft which don't, the number of engines etc. 

If you want users to be able to drop-in/select pretty much any JSBSim FDM as opposed to you only providing a handful that you can setup and control then I suggest you take a look at how FlightGear manage this.

**Comment by GloryENG** [comment]
Thanks so so much @seanmcleod @dpculp 

I think I should take a look and study "FlightGear.cxx"

you mean, the easiest way to communicate or interface between out-running simulator and JSBsim is look and modify FlightGear.cxx, right?

Is there any other things good to know for my case?

Furthermore, I do surely and clearly find and check the accessible properties, parameters that user can get and set during JSBSim simulation. How do I do?

**Reply by dpculp** [reply]
I must admit I still don't have a clear picture of how your simulation is structured.  Does it look like this, or different?

![1](https://user-images.githubusercontent.com/21369577/210669215-e0f668bc-4bb5-4837-8dce-25136f13cb5e.jpeg)

**Reply by GloryENG** [reply]
Yes, as you described perfectly.
So, instead of flightgear, my own simulation viewer will shows the 3d plane, and at the same time, user would set the initial condition, weather, as well as ordinary flight control command. In this point, I am considering how to get and set the properties(?) parameters(?) in JSBSim.

**Reply by dpculp** [reply]
OK, so if I understand correctly, you are going to run JSBSim as a separate process, then you communicate with it as shown in the diagram?

How many separate apps are you using?

Which app reads the keyboard and joystick?  Is it the same app as the viewer?

**Reply by seanmcleod** [reply]
> I need to write the code regarding copy_to_JSBSim and copy_from_JSBSIm...

@dpculp my understanding is that he will have a single process linking in JSBSim and communicating with JSBSim in-process via `copy_to_JSBSim()` and `copy_from_JSBSim()` and that this process will be a Unity based app. So he'll write code, or make use of possibly some Unity APIs to poll joysticks, keyboards etc. then pass that to JSBSim via `copy_to_JSBSim()` and retrieve state information like attitude, position etc. from `copy_from_JSBSim()` and use that to get his Unity app to render an outside view etc.

So pretty much re-implementing FlightGear.

**Comment by seanmcleod** [comment]
I'm not actually familiar with FlightGear's source code, my suggestion to look at how FlightGear does things was based on the fact that you want to be able to handle any JSBSim aircraft model in your sim and that's what FlightGear does. So take a look at how they handle things in terms of different engine types, different number of engines, different sets of controls etc.

The `FlightGear.cxx` in the JSBSim may be a good starting point, but you may need to head over to the FlightGear repo and take a more detailed look at their developer documentation, source code etc.

**Comment by pd3a** [comment]
Hi, did you figure out a way for this? To use JSBSim output to visualize 3d rendering in unity?

**Comment by jonbarril** [comment]
I too am interested in using JSBSim in Unity for flight dynamics. Has anyone achieved this?
